src/Featurizers/ForwardFillImputerFeaturizer.h

```diff
@@ -0,0 +1,88 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
+
+#include "../Archive.h"
+#include "../Traits.h"
+
+#include "Components/InferenceOnlyFeaturizerImpl.h"
+
+namespace Microsoft {
+namespace Featurizer {
+namespace Featurizers {
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         ForwardFillImputerTransformer
+///  \brief         Imputer that populates null incoming values with the last
+///                 non-null value encountered.
+///
+template <typename T>
+class ForwardFillImputerTransformer :
+    public Components::InferenceOnlyTransformerImpl<
+        typename Traits<T>::nullable_type,
+        T
+    > {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    static_assert(Traits<T>::IsNullableType == false || Traits<T>::IsNativeNullableType, "'T' must not be a nullable type");
+
+    using BaseType =
+        Components::InferenceOnlyTransformerImpl<
+            typename Traits<T>::nullable_type,
+            T
+        >;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    T                                       _lastValue;
+    bool                                    _validLastValue = false;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC runs into problems when the declaration and definition are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        // ----------------------------------------------------------------------
+        using TheseTraits                   = Traits<typename BaseType::InputType>;
+        // ----------------------------------------------------------------------
+
+        if(TheseTraits::IsNull(input)) {
+            if(_validLastValue == false)
+                throw std::runtime_error("No source value for forward fill");

```

Comment:
>throw std::runtime_error("No source value for forward fill"); [](start = 16, length = 61)

Who is supposed to handle this or deal with this? Is this a caller problem or a data problem?

Reply:
This would be a data problem; it indicates that a null value has been received before a valid value, and there is nothing that can be used to populate this value. I considered avoiding the error case by making all output types optional (which would let us propagate null values for this particular case), but that seemed overly heavy. If we find that this is a common case, we can introduce more intelligent logic if the output happens to be a nullable type.

I'm going to add some comments in the code to describe this scenario.

---
In reply to: [377934541](https://github.com/microsoft/FeaturizersLibrary/pull/139#discussion_r377934541) [](ancestors = 377934541)

Reply:
Actually, now that I think about it, I'm going to add a constructor that let's a caller optionally provide a default value so that this scenario will be less common.

---
In reply to: [379533991](https://github.com/microsoft/FeaturizersLibrary/pull/139#discussion_r379533991) [](ancestors = 379533991,377934541)